File: app/src/main/java/com/aspirasoft/huntrek/ActivitySplash.java
package com.aspirasoft.huntrek;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.view.View;

import com.aspirasoft.huntrek.utils.Database;

import java.util.ArrayList;
import java.util.Timer;
import java.util.TimerTask;

public class ActivitySplash extends Activity {

    private static final int LOCATION_REQUEST_CODE = 200;
    ArrayList<String> locationPermissions = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        View decorView = getWindow().getDecorView();
        int uiOptions = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LAYOUT_STABLE;
        decorView.setSystemUiVisibility(uiOptions);
        setContentView(R.layout.activity_splash);
        Database.init(this);

        locationPermissions.add(Manifest.permission.ACCESS_FINE_LOCATION);
        locationPermissions.add(Manifest.permission.ACCESS_COARSE_LOCATION);

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, locationPermissions.toArray(new String[]{}), LOCATION_REQUEST_CODE);
        } else {
            new Timer().schedule(new TimerTask() {
                @Override
                public void run() {
                    startActivity(new Intent(getApplicationContext(), ActivityHunt.class));
                    finish();
                }
            }, 1500L);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        if (requestCode == LOCATION_REQUEST_CODE) {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
                    ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                new Timer().schedule(new TimerTask() {
                    @Override
                    public void run() {
                        startActivity(new Intent(getApplicationContext(), ActivityHunt.class));
                        finish();
                    }
                }, 1500L);
            } else {
                finish();
            }
        }
    }
}

File: app/src/main/java/dev/aspirasoft/huntit/ui/activity/ActivitySplash.kt
package dev.aspirasoft.huntit.ui.activity

import android.Manifest
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import com.google.firebase.auth.FirebaseAuth
import dev.aspirasoft.huntit.R
import dev.aspirasoft.huntit.data.repo.AuthRepository
import dev.aspirasoft.huntit.model.AuthStatus
import dev.aspirasoft.huntit.utils.LocationPermissionUtil
import java.util.*


class ActivitySplash : FullScreenActivity() {

    private lateinit var repo: AuthRepository

    private val mGameLaunchTask = GameLaunchTask()

    private val permissions = arrayOf(
        Manifest.permission.ACCESS_FINE_LOCATION,
        Manifest.permission.ACCESS_COARSE_LOCATION
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash)

        repo = AuthRepository()
    }

    override fun onResume() {
        super.onResume()
        // ARCoreUtil.checkAndInstall(this)
        when (LocationPermissionUtil.hasLocationPermissions(this)) {
            true -> mGameLaunchTask.startDelayed()
            false -> LocationPermissionUtil.askLocationPermissions(this)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            LocationPermissionUtil.LOCATION_PERMISSION_CODE -> {
                when (LocationPermissionUtil.hasLocationPermissions(this)) {
                    true -> mGameLaunchTask.startDelayed()
                    false -> onLocationPermissionDenied()
                }
            }
        }
    }

    private fun onLocationPermissionDenied() {
        Toast.makeText(this,
            "Location permission is needed to run this application",
            Toast.LENGTH_LONG).show()
        // todo: provide a way to turn on location permissions
        finish()
    }

    private inner class GameLaunchTask : TimerTask() {

        private fun proceed(status: AuthStatus) {
            startActivity(Intent(applicationContext, when (status) {
                AuthStatus.SIGNED_IN -> ActivityHunt::class.java
                AuthStatus.SIGNUP_INCOMPLETE -> ActivitySignUp::class.java
                else -> ActivitySignIn::class.java
            }))
            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
            finish()
        }

        override fun run() {
            when (repo.isSignedIn) {
                true -> proceed(AuthStatus.SIGNED_IN)
                else -> when (FirebaseAuth.getInstance().currentUser) {
                    null -> proceed(AuthStatus.SIGNED_OUT)
                    else -> proceed(AuthStatus.SIGNUP_INCOMPLETE)
                }
            }
        }

        fun startDelayed() {
            Timer().schedule(this, 1500L)
        }

    }

    companion object {
        private const val LOCATION_REQUEST_CODE = 200
    }

}